scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -301,6 +317,45 @@ static struct cpumask *lookup_layer_cpumask(int idx)
 	}
 }
 
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, int);
+	__type(value, struct lock_wrapper);
+	__uint(max_entries, LO_FALLBACK_DSQ);
+	__uint(map_flags, 0);
+} dsq_ctx_locks SEC(".maps");
+
+static void calc_dsq_usage(u64 dsq_id)
+{
+	struct dsq_ctx *dsqc;
+	struct lock_wrapper *lockw;
+
+	if (!(dsqc = lookup_dsq_ctx(dsq_id)) ||
+	    !(lockw = bpf_map_lookup_elem(&dsq_ctx_locks, &dsq_id)))
+		return;
+
+	bpf_spin_lock(&lockw->lock);
+	dsqc->usage_avg = (dsqc->usage_avg / 8) * 7;
```

Comment:
@htejun I'm not convinced this is better than using `ravg_accumulate`, but wasn't sure if it is ok to call it during stopping or if it should be called periodically.

Reply:
`ravg` can be used here. It doesn't need periodic calls but it does need timestamp on each accumulation and when the value is read. Here, timestamp is already acquired anyway, so shouldn't add much overhead.

Reply:
But why does it need to average out usage? Can't it just track absoulate usage and control accordingly?

Reply:
We were concerned that tracking absolute usage over the lifetime of the scheduler would cause massive priority inflation for layers with moderate weights that go unused for a while, in that we'd end up seriously indebted to these processes in the long term because we didn't serve them when they had no tasks queued. Looking at usage with a big recency bias seemed to solve this case nicely, while still helping processes that would get less than their weight in the short term. Would be interested in more of your thoughts!

Reply:
The usual strategy to handle that is capping the maximum deltas. e.g. Imagine putting the layers on a vtime line like we do with tasks, the reason why task vtime lines enforce min_vtime when a task wakes up is for the same reason, as the negative delta represents the budget it has accumulated while it wasn't running and we don't want it to dominate. This way of expressing it in vtime scheme maps closely to maximum accumulated token limit on token based distribution schemes.

Using running avg also does limit the amount of deviation but as an almost secondary effect. It will be conceptually cleaner and easier to modulate if the implemented mechanism directly addresses the issue (here, accumulating too much budget delta).